ext_authz: Add ability to enable via a runtime fractional percent

gkleiman opened the pull request:
Added a `filter_enabled` field to the HTTP external authorization filter, which makes it possible to enable the filter for just a percentage of requests. This field can be configured via runtime.

Risk Level: Medium
Testing: Unit tests, manual testing modifying the new field via a config file and via the `POST /runtime_modify` admin endpoint
Docs Changes: Updated ext_authz_filter.rst and added inline proto documentation for the new field
Release Notes: ext_authz: added `filter_enabled` RuntimeFractionalPercent flag to filter

Signed-off-by: Gastón Kleiman <gkleiman@lyft.com>

wgallagher commented on test/extensions/filters/http/ext_authz/ext_authz_test.cc:
```diff
@@ -830,6 +830,57 @@ TEST_F(HttpFilterTest, MetadataContext) {
+  EXPECT_EQ(Http::FilterHeadersStatus::Continue, filter_->decodeHeaders(request_headers_, false));
+}
+
+// Test that filter can be disabled via the filter_enabled field.
```
typo s/disabled/enabled/

gkleiman replied:
Good catch! Copypasta always gets me 😳

wgallagher commented on test/extensions/filters/http/ext_authz/ext_authz_test.cc:
```diff
@@ -830,6 +830,57 @@ TEST_F(HttpFilterTest, MetadataContext) {
+      denominator: HUNDRED
+  )EOF");
+
+  ON_CALL(runtime_.snapshot_,
```
maybe add prepareCheck here, too?

gkleiman replied:
Good idea. I was actually on the fence, because I usually hate copying and pasting boilerplate that is not needed.

However in this case, without `prepareCheck()` the test will segfault if a check is called. That isn't very nice experience, so I will add the call =).

gkleiman replied:
Actually... I can't add `prepareCheck()` as-is, because it will add an expectation that will only be satisfied if `check()` is called.... and this test verifies that `check()` is NOT called.

gkleiman commented on docs/root/configuration/http/http_filters/ext_authz_filter.rst:
```diff
@@ -131,3 +131,21 @@ The HTTP filter outputs statistics in the *cluster.<route target cluster>.ext_au
+
+The external authorization filter supports the following RuntimeFractionalPercent setting:
+
+filter_enabled
```
I followed the docs pattern used for the `RuntimeFractionalPercent` settings of for the [CSRF](https://github.com/envoyproxy/envoy/blob/92c085afaeeed138067f289fa9294d56afc91e25/docs/root/configuration/http/http_filters/csrf_filter.rst#runtime) and [CORS](https://github.com/envoyproxy/envoy/blob/master/docs/root/configuration/http/http_filters/cors_filter.rst#runtime) filters.

Should I keep this as-is or follow up with a PR that also updates those docs?

mattklein123 replied:
It's not a big deal but I would probably just fix them all in this PR if you feel like it.

gkleiman replied:
I updated the format used for the flag I'm introducing. The filter doc page has now fewer information and ref links the field for extra info.

The CSRF and CORS plugins actually follow the opposite approach — the filter doc page has most of the details and the field doc has just a brief description with a ref link to the filter doc page.

Please let me know which approach you like best and I'll try to make these three cases consistent =).

mattklein123 replied:
Personally I like what you did here, I think it's cleaner to read, but I don't feel strongly about it. I would just make it consistent whichever way you choose. Thank you! :)

mattklein123 commented on source/extensions/filters/http/ext_authz/ext_authz.h:
```diff
@@ -45,8 +45,14 @@ class FilterConfig {
+        filter_enabled_runtime_key_(config.filter_enabled().runtime_key()),
+        filter_enabled_default_value_(config.has_filter_enabled()
+                                          ? absl::optional<envoy::type::FractionalPercent>(
+                                                config.filter_enabled().default_value())
```
Oops sorry disregard this.

gkleiman replied:
It is a bit confusing, but the [proto field](https://github.com/envoyproxy/envoy/blob/986173ed516dcc1c3dea7db90659ed993d0aad75/api/envoy/api/v2/core/base.proto#L293) with the real value is called `default_value`.

I don't have a config loading test, but the tests [[1]](https://github.com/gkleiman/envoy/blob/46df1acf0c585d51ac0c0c018393676811f53cc9/test/extensions/filters/http/ext_authz/ext_authz_test.cc#L846-L849) [[2]](https://github.com/gkleiman/envoy/blob/46df1acf0c585d51ac0c0c018393676811f53cc9/test/extensions/filters/http/ext_authz/ext_authz_test.cc#L872-L875) I added do verify that the runtime loader is called with the right value.

Writing a config loading test might be a bit tricky in this case, because the config values are stored in private fields and not directly exposed.

Please re-open this if I you still think I'm using the wrong value or if I haven't convinced you that the current tests are enough =).

mattklein123 replied:
Sorry see above, disregard my original comment. This is correct.

gkleiman commented on source/extensions/filters/http/ext_authz/ext_authz.h:
```diff
@@ -95,6 +110,8 @@ class FilterConfig {
   const bool clear_route_cache_;
   const uint32_t max_request_bytes_;
   const Http::Code status_on_error_;
+  const std::string filter_enabled_runtime_key_;
```
Grepping for `runtime_key_` revealed that this patterns is indeed pretty common.

I'd be happy to follow up to this PR with another one that adds a new helper `Runtime::FractionalPercent` class and updates those places to use it.

mattklein123 replied:
I would just add the utility now and use it in this PR, and then if you could do a follow up for other places that would be great.

gkleiman commented:
@mattklein123 thanks for the prompt reviews!

I'll work on the utility class and update the PR on Monday.

In the meantime I would appreciate if you could take a look at the updated docs [[1]](https://github.com/envoyproxy/envoy/pull/8763#discussion_r339261652) to see whether I should also update the CSRF/CORS docs.

mattklein123 commented on docs/root/configuration/http/http_filters/ext_authz_filter.rst:
```diff
@@ -131,3 +131,12 @@ The HTTP filter outputs statistics in the *cluster.<route target cluster>.ext_au
+
+Runtime
+-------
+The fraction of requests for which the filter is enabled can be configured via the the
```
typo "the the"

mattklein123 commented on docs/root/configuration/http/http_filters/ext_authz_filter.rst:
```diff
@@ -131,3 +131,12 @@ The HTTP filter outputs statistics in the *cluster.<route target cluster>.ext_au
+:ref:`runtime_key <envoy_api_msg_core.runtimefractionalpercent>` value of the :ref:`filter_enabled
+<envoy_api_msg_config.filter.http.ext_authz.v2.ExtAuthz>` field.
+
+To determine if the filter is enabled you can check the runtime values via the admin panel at
```
nit: I would probably kill this as we don't do this elsewhere, but not a big deal.

mattklein123 commented on source/common/runtime/runtime_features.h:
```diff
@@ -56,5 +56,21 @@ class FeatureFlag {
+// Helper class for runtime-derived fractional percent flags.
+class FractionalPercent {
+public:
+  FractionalPercent(const envoy::api::v2::core::RuntimeFractionalPercent fractional_percent_proto,
```
nit: pass by const ref (I would fix above also)

mattklein123 commented on source/extensions/filters/http/ext_authz/ext_authz.h:
```diff
@@ -46,7 +47,12 @@ class FilterConfig {
         status_on_error_(toErrorCode(config.status_on_error().code())), local_info_(local_info),
-        scope_(scope), runtime_(runtime), http_context_(http_context), pool_(scope.symbolTable()),
+        scope_(scope), runtime_(runtime), http_context_(http_context),
+        filter_enabled_(config.has_filter_enabled()
```
nit: optimally, the optional would be embedded inside the helper. One way of doing this would be to either pass a pointer to the field or nullptr, and then do the logic inside such that if nullptr it always returns true?

gkleiman replied:
I considered this while implementing the helper, but decided to go this route, because there are some cases for which the logic inside should return `false` if it's a `nullptr`, e.g., for `shadow_enabled`.

gkleiman commented:
@mattklein123 thanks for your reviews.

I've merged `master`, addressed your feedback, and added some more doc improvements. I think (hope?) it's good to go.

I'll follow up with an issue to update the other filters to use `Runtime::FractionalPercent`.

mattklein123 commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -475,36 +475,28 @@ message CorsPolicy {
     google.protobuf.BoolValue enabled = 7 [deprecated = true];
 
-    // Specifies if CORS is enabled.
+    // Specifies the % of requests for which the CORS filter is enabled.
```
Sorry I didn't catch this before, but I think these docs should still say that the default if not specified is 100%? Same elsewhere?

gkleiman replied:
👍

mattklein123 commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -82,6 +82,21 @@ message ExtAuthz {
+  //
+  // If this field is not specified, the filter will be enabled for all requests.
+  //
+  // Parsing this field is implemented such that the runtime key's data may be represented
```
Can we move this common doc into the docs from runtime fractional percent?

mattklein123 commented:
Sorry can you merge master to fix format.

/wait

gkleiman commented:
`master` merged =).
